### PR TITLE
Flexible property prediction heads

### PR DESCRIPTION
**Pull request by M-R-Schaefer**
I have added property prediction heads to the EnergyModel.
These can predict per-atom or per-structure quantities with either l0, l1 or symmetric traceless l2 symmetry.

Further, I have added a LatentEwald module which gives us access to long range interactions. It requires a property head which predicts charges.

The datapipeline can now read in arbitrary properties from h5 files. The keywords in the loss function determine which properties are read.

other changes
- slight cleanup of the dtype configs
- moved atomistic model functionality directly into the energy model

### CODE REVIEW

File: apax/nn/models.py [hunk 1]
```diff
@@ -186,28 +180,19 @@ def mean_energy_fn(
...
-    return grad_i_fn
-
-
 def make_member_chunk_jac(energy_model, start, end):
     def energy_chunk_fn(R, Z, neighbor, box, offsets):
         Ei = energy_model(R, Z, neighbor, box, offsets)[start:end]
         return Ei
 
-    grad_i_fn = jax.jacrev(energy_chunk_fn)
+    grad_i_fn = jax.jacrev(energy_chunk_fn)  # TODO
```
Comment (PythonFZ) [inline apax/nn/models.py:195]: TODO?

File: apax/utils/data.py [hunk 2]
```diff
@@ -21,28 +21,7 @@ def make_minimal_input():
...
-        np.savez(
-            "data_path_labels.npz",
-            dipole=dipole,
-            charge=charge,
-            mat=mat,
-            shape=shape,
-        )
-
-    shape has to be in the same order than the parameters
+    """
```
Comment (PythonFZ) [inline apax/utils/data.py:24]: What happend here?

**Review by PythonFZ — APPROVED** [review]
Looks good - I left a few questions but no real requests for review. I think this can be merged but I have not tested it locally myself.

File: apax/bal/api.py [hunk 3]
```diff
@@ -150,12 +151,13 @@ def kernel_selection(
     _, init_box = dataset.init_input()
 
     Builder = config.model.get_builder()
-    builder = Builder(config.model.get_dict(), n_species=119)
+    builder = Builder(config.model.model_dump(), n_species=119)
 
-    model = builder.build_energy_model(apply_mask=True, init_box=init_box)
+    energy_model = builder.build_energy_model(apply_mask=True, init_box=init_box)
+    energy_model = make_energy_only_model(energy_model.apply)
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/bal/api.py:157]: yes, build_energy_model gives a model for energy and a dict of additional properties. Some tasks require a model that specifically only returns a scalar energy (no properties, no ensemble E preds)

File: apax/data/input_pipeline.py [hunk 4]
```diff
@@ -90,6 +90,7 @@ def __init__(
         n_jit_steps=1,
         pos_unit: str = "Ang",
         energy_unit: str = "eV",
+        additional_properties: list[tuple] = [],
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/data/input_pipeline.py:93]: what's wrong with it? No particular reason, could be iterable iterable

File: apax/data/input_pipeline.py [hunk 5]
```diff
@@ -161,8 +165,16 @@ def prepare_data(self, i):
             labels["forces"] = np.pad(
                 labels["forces"], ((0, zeros_to_add), (0, 0)), "constant"
             )
+
+        for prop in self.additional_properties:
+            name, shape = prop
+            if shape[0] == "natoms":
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/data/input_pipeline.py:171]: it could be natoms or nneighbors

File: apax/layers/descriptor/basis_functions.py [hunk 6]
```diff
@@ -69,10 +73,9 @@ class RadialFunction(nn.Module):
     dtype: Any = jnp.float32
 
     def setup(self):
+        dtype = str_to_dtype(self.dtype)
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/layers/descriptor/basis_functions.py:76]: self.dtype is a string e.g. "fp32" but gets converted to the proper dtype by that function. it's not great but handling this via pydantic validators caused a lot of problems.
Under jit, it is only converted once, even if used in __call__

File: apax/layers/properties.py [hunk 7]
```diff
@@ -35,3 +40,87 @@ def U(eps):
...
     dUdV = jax.grad(U)
     return dUdV(zero)
+
+
+class PropertyHead(nn.Module):
+    """
+    the readout is currently limited to a single number
+    """
+
+    pname: str
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/layers/properties.py:50]: name is a protected attribute by flax. We can't change it unfortunately

File: apax/layers/readout.py [hunk 8]
```diff
@@ -43,4 +46,5 @@ def setup(self):
 
     def __call__(self, x):
         h = self.sequential(x)
+        # TODO should we move aggregation here?
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/layers/readout.py:49]: more of a question to myself :sweat_smile:

File: apax/layers/properties.py [hunk 9]
```diff
@@ -35,3 +40,87 @@ def U(eps):
...
+            # ensemble detected
+            is_ensemble = True
+            n_ens = jnp.size(h, axis=1)
+            h = h[..., None]
+            h = jnp.transpose(h, (1, 0, 2))
+
+        p_i = h * self.scale[Z] + self.shift_param[Z]
+
+        if self.mode == "l0":
+            p_i = p_i
```
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/layers/properties.py:83]: can be removed

File: apax/utils/data.py (hunk 2 above)
Comment (M-R-Schaefer) [reply apax/utils/data.py:24]: the docstring referenced an old unused method where additional data was loaded from an npz file.

File: apax/data/input_pipeline.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (M-R-Schaefer) [reply apax/data/input_pipeline.py:93]: I don't think this matters all that much. I like having loops that get skipped for empty lists. If I set it to None I have to include a conditional in the class.
